File: project/6.10_ObjectクラスとClassクラス/java/java_ex6_10_1/src/java_ex6_10_1/Ex6_10_1.java
package java_ex6_10_1;

class Person {
	static private int num = 0;
	String name;
	
	Person( String name ) {
		num++;
		this.name = name;
	}
	
	public String toString() {
		return "num=" + num + ", name=" + this.name;
	}
}


public class Ex6_10_1 {

	public static void main(String[] args) {
		Person p1 = new Person( "Keroichi" );
		System.out.println(p1);
		
		Person p2 = new Person( "Kerokero" );
		System.out.println(p2);

		System.out.println(p1);
	}

}

File: project/6.10_ObjectクラスとClassクラス/py/py_ex6_10_1/src/ex6_10_1.py
'''
Created on 2023/05/11

@author: Keroichi.T
'''

import inspect

class Person( object ):
    num = 0

    def __init__(self, name):
        # クラス変数へのアクセス
        Person.num += 1
        self.name = name
    
    # objectのメソッドstrをオーバーライド
    def __str__(self)->str:
        return f"num={Person.num}, name={self.name}"

def main():
    
    p1 = Person( "Keroichi" )
    print( p1 )
    
    p2 = Person( "Kerokero" )
    print( p2 )
    
    print( p1 )
    
    print( hasattr(Person, "__str__"))
    print( hasattr(Person, "__add__"))

    # オブジェクトが持つメソッドの一覧を取得する
    for m in inspect.getmembers(object()):
        print( m )

if __name__ == '__main__':
    main()